sdk/communication/communication-common/src/entraTokenCredential.ts

```diff
@@ -0,0 +1,136 @@
+   * The Entra ID token credential.
+   */
+  tokenCredential: TokenCredential;
+  /**
+   * The scopes for retrieving the Entra ID access token.
+   */
+  scopes: string[];
+}
+
+/**
+ * EntraTokenCredential
+ */
+export class EntraTokenCredential implements AcsTokenCredential {
+  private isPending: Promise<AccessToken> | null;
+  private result = {
+    entraToken: undefined as string | undefined,
+    acsToken: { token: "", expiresOnTimestamp: 0 },
+  };
+  private client: Client;
+  private httpClient: HttpClient;
+
+  constructor(private options: EntraCommunicationTokenCredentialOptions) {
+    this.client = getClient(options.resourceEndpoint);
+    this.httpClient = createDefaultHttpClient();
+
+    // immediately fetch the token to pre-warm
+    this.isPending = this.getToken();
+  }
+
+  public async getToken(options?: CommunicationGetTokenOptions): Promise<AccessToken> {
+    if (options?.abortSignal?.aborted) {
+      return { token: "", expiresOnTimestamp: 0 };
+    }
+
+    // we're awaiting the token fetch, so we don't want to start another one
+    // however, we're ignoring the new abortSignal, unfortunately
+    if (!this.isPending) {
+      this.isPending = this.getTokenInternal(options);
+    }
+
+    try {
+      await this.isPending;
+    } finally {
+      this.isPending = null;
+    }
+
+    return this.result.acsToken;
+  }
+
+  private async getTokenInternal(options?: CommunicationGetTokenOptions): Promise<AccessToken> {
+    const getTokenOptions = options?.abortSignal ? { abortSignal: options.abortSignal } : undefined;
+    const token = await this.options.tokenCredential.getToken(this.options.scopes, getTokenOptions);
+
+    if (token === null) {
+      this.result = {
+        entraToken: undefined,
+        acsToken: { token: "", expiresOnTimestamp: 0 },
+      };
+    } else if (this.result.acsToken.token === "" || token.token !== this.result.entraToken) {
+      const acsToken = await this.exchangeEntraToken(
+        this.options.resourceEndpoint,
+        token.token,
+        getTokenOptions,
+      );
+      this.result = {
+        entraToken: token.token,
+        acsToken,
+      };
+    }
+
+    return this.result.acsToken;
+  }
+
+  public dispose(): void {
+    /* intentionally empty */
+  }
+
+  private async exchangeEntraToken(
+    resourceEndpoint: string,
+    entraToken: string,
+    options?: { abortSignal: AbortSignalLike },
+  ): Promise<AccessToken> {
+    const request = createPipelineRequest({
+      url: `${resourceEndpoint}/access/entra:exchangeToken`,
+      method: "POST",
+      headers: createHttpHeaders({
+        Authorization: `Bearer ${entraToken}`,
+      }),
+      abortSignal: options?.abortSignal,
+    });
+    const response = await this.client.pipeline.sendRequest(this.httpClient, request);
+
+    if (response.status !== 200 || !response.bodyAsText) {
+      throw new Error(
+        `Failed to exchange entra token. Status: ${response.status}, Body: ${response.bodyAsText}`,
```

Comment:
should the clients know about the token exchange? Maybe can just provide a generic exception that the request failed